pkg/tagger/tagstore/entity_tags.go

```diff
@@ -70,64 +70,64 @@ func (e *EntityTags) toEntity() types.Entity {
 	}
 }
 
-type tagPriority struct {
-	tag         string                       // full tag
-	priority    collectors.CollectorPriority // collector priority
-	cardinality collectors.TagCardinality    // cardinality level of the tag (low, orchestrator, high)
-}
-
 func (e *EntityTags) computeCache() {
 	if e.cacheValid {
 		return
 	}
 
-	var sources []string
-	tagPrioMapper := make(map[string][]tagPriority)
+	tagList := make(map[collectors.TagCardinality][]string)
+	tagMap := make(map[string]collectors.CollectorPriority)
 
-	for source, tags := range e.sourceTags {
+	var sources []string
+	for source := range e.sourceTags {
 		sources = append(sources, source)
-		insertWithPriority(tagPrioMapper, tags.lowCardTags, source, collectors.LowCardinality)
-		insertWithPriority(tagPrioMapper, tags.orchestratorCardTags, source, collectors.OrchestratorCardinality)
-		insertWithPriority(tagPrioMapper, tags.highCardTags, source, collectors.HighCardinality)
 	}
 
-	var lowCardTags []string
-	var orchestratorCardTags []string
-	var highCardTags []string
-	for _, tags := range tagPrioMapper {
-		for i := 0; i < len(tags); i++ {
-			insert := true
-			for j := 0; j < len(tags); j++ {
-				// if we find a duplicate tag with higher priority we do not insert the tag
-				if i != j && tags[i].priority < tags[j].priority {
-					insert = false
-					break
-				}
-			}
-			if !insert {
+	// sort sources in descending order of priority. assumes lowest if
+	// priority is not declared
+	sort.Slice(sources, func(i, j int) bool {
+		sourceI := sources[i]
+		sourceJ := sources[j]
+		return collectors.CollectorPriorities[sourceI] > collectors.CollectorPriorities[sourceJ]
+	})
+
+	// insertWithPriority prevents two collectors of different priorities
+	// from reporting duplicated tags. two collectors of the same priority
+	// still can report duplicated tags.
```

Comment:
In terms of which tag "wins" in case of two collectors of different priorities, it's the one of the collector with the higher priority, whichever cardinality that reports.

You can also read "can report duplicated tags" as "they'll be duplicated, but really collectors shouldn't do it". I'll update the comment to make both points clearer.